### DISCUSSION TITLE
Run Custom Docker Image of Caprover

### DISCUSSION THREAD

**Discussion by KellerKev** [caprover/caprover#1874] [Q&A]
I want to run a custom caprover image where I changed some code, build the new image and then I run it. 
But every time I run it, my code changes are not applied and the image name that I see running is caprover/caprover:1.11.0 instead of my custom image name. 

What am I doing wrong? 

Steps:

1. Git clone caprover
2. Copy dockerfile-captain.release to Dockerfile
3. docker builder prune
4. docker build . --no-cache
5. docker image XXX tag keller/cap:latest
6. docker run -p 80:80 -p 443:443 -p 3000:3000 -v /var/run/docker.sock:/var/run/docker.sock -v /captain:/captain -e ACCEPTED_TERMS=true keller/cap

**Comment by githubsaturn** [comment]
Look into captainconstants.ts file and update the image name

**Reply by KellerKev** [reply]
That did not help.. 

const CONSTANT_FILE_OVERRIDE_USER =
    CAPTAIN_DATA_DIRECTORY + '/config-override.json'

const configs = {
    publishedNameOnDockerHub: 'keller/cap',

    version: 'latest',

**Reply by KellerKev** [reply]
When I first run the container it looks ok..






 CONTAINER ID   IMAGE               COMMAND                  CREATED          STATUS          PORTS                                                                                                                 NAMES
93198408129d   keller/cap:latest   "docker-entrypoint.s…"   23 seconds ago   Up 21 seconds   0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp, 0.0.0.0

**Reply by githubsaturn** [reply]
Very likely you're not committing your changes. That's the only place where the image is being referenced.

You can also look at how edge overrides the image: https://github.com/caprover/caprover/blob/master/edge-override.json

https://github.com/caprover/caprover/blob/80cb3a3a90ba5840c020a7e3ddaf00cc19b239b4/dockerfile-captain.edge#L15

**Comment by KellerKev** [comment]
but once caprover is fully up and running we are back to the vanilla version: 

CONTAINER ID   IMAGE                              COMMAND                  CREATED          STATUS          PORTS                                                                      NAMES
9d3423561d2b   nginx:1                            "/docker-entrypoint.…"   5 seconds ago    Up 3 seconds    0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp   captain-nginx.1.91j3hhy38bca85rle45r4qmf2
78d123a0a411   caprover/certbot-sleeping:v1.6.0   "/bin/sh -c 'sleep 9…"   19 seconds ago   Up 16 seconds   80/tcp, 443/tcp                                                            captain-certbot.1.fayjzc1ljfar5veggfuoceair
60c28bd4994c   caprover/caprover:1.11.0           "docker-entrypoint.s…"   37 seconds ago   Up 35 seconds   0.0.0.0:3000->3000/tcp, :::3000->3000/tcp                                  captain-captain.1.zse1eibgtlm9yiqsjhswv2igx

**Comment by KellerKev** [comment]
I suppoise I its expectign it on Dockerhub instead of using the local image:  

return DockerApi.get().createServiceOnNodeId(
                captainNameAndVersion,
                CaptainConstants.captainServiceName,
                ports,
                nodeId,
                volumeToMount,
                env,

**Comment by KellerKev** [comment]
Is there a way to just use the local image without pulling another one or to use a private registry?

**Reply by githubsaturn** [reply]
That is definitely doable. That's how we build and develop CapRover locally.

https://github.com/caprover/caprover/blob/master/CONTRIBUTING.md#running-backend-dev-environment
